src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-utils.c

```diff
@@ -804,46 +842,117 @@ int xclmgmt_program_shell(struct xclmgmt_dev *lro)
 
 }
 
-int xclmgmt_load_fdt(struct xclmgmt_dev *lro)
+static int xclmgmt_refresh_fdt_blob(struct xclmgmt_dev *lro, const char *fw_buf)
 {
-	struct xocl_board_private *dev_info = &lro->core.priv;
-	const struct axlf_section_header	*dtc_header;
-	struct axlf				*bin_axlf;
-	int					ret;
-	char					*fw_buf = NULL;
-	size_t					fw_size = 0;
-
-
-	if (xocl_subdev_is_vsec_recovery(lro)) {
-		mgmt_info(lro, "Skip load_fdt for vsec Golden image");
-		(void) xocl_peer_listen(lro, xclmgmt_mailbox_srv, (void *)lro);
-		return 0;
-	}
+	const struct axlf_section_header	*dtc_header = NULL;
+	struct axlf				*bin_axlf = NULL;
+	int					ret = 0;
 
-	mutex_lock(&lro->busy_mutex);
-	ret = xocl_rom_load_firmware(lro, &fw_buf, &fw_size);
-	if (ret) {
-		mgmt_err(lro, "ROM firmware load failure %d", ret);
-		goto failed;
-	}
 	bin_axlf = (struct axlf *)fw_buf;
 
 	dtc_header = xocl_axlf_section_header(lro, bin_axlf, PARTITION_METADATA);
 	if (!dtc_header) {
-		ret = -ENOENT;
 		mgmt_err(lro, "Firmware does not contain PARTITION_METADATA");
-		goto failed;
+		return -ENOENT;
 	}
 
 	ret = xocl_fdt_blob_input(lro,
 			(char *)fw_buf + dtc_header->m_sectionOffset,
 			dtc_header->m_sectionSize, XOCL_SUBDEV_LEVEL_BLD,
 			bin_axlf->m_header.m_platformVBNV);
-	if (ret) {
+	if (ret)
 		mgmt_err(lro, "Invalid PARTITION_METADATA");
-		goto failed;
+
+	xclmgmt_debug_dump_uuid(lro);
+
+	return ret;
+}
+
+/*
+ * load firmware, the fdt blob in the partition metadata, and
+ * refresh cached fdt_blob
+ */
+static int xclmgmt_reload_fdt_blob(struct xclmgmt_dev *lro)
+{
+	char	*fw_buf = NULL;
+	size_t	fw_size = 0;
+	int	ret = 0;
+
+	BUG_ON(!mutex_is_locked(&lro->busy_mutex));
+
+	ret = xocl_rom_load_firmware(lro, &fw_buf, &fw_size);
+	if (ret) {
+		mgmt_err(lro, "ROM firmware load failure %d", ret);
+		return ret;
+	}
+
+	ret = xclmgmt_refresh_fdt_blob(lro, fw_buf);
+
+	vfree(fw_buf);
+	return ret;
+}
+
+/*
+ * load firmware from vmr devices
+ *   - then clean up existing fdt_blob if this is a hot reset request
+ *   - after fdt_blob is refreshed, the blp_blob should be updated.
+ */
+static int xclmgmt_reload_fdt_blob_vmr(struct xclmgmt_dev *lro)
+{
+	char	*fw_buf = NULL;
+	size_t	fw_size = 0;
+	int	ret = 0;
+
+	BUG_ON(!mutex_is_locked(&lro->busy_mutex));
+
+	ret = xocl_vmr_load_firmware(lro, &fw_buf, &fw_size);
+	if (ret)
+		return ret;
+
+	/* clean up existing fdt_blob */
+	if (lro->core.fdt_blob) {
+		vfree(lro->core.fdt_blob);
+		lro->core.fdt_blob = NULL;
+	}
+
+	ret = xclmgmt_refresh_fdt_blob(lro, fw_buf);
+
+	/* replace blp_blob */
+	if (lro->core.blp_blob) {
+		vfree(lro->core.blp_blob);
+		lro->core.blp_blob = vmalloc(fdt_totalsize(lro->core.fdt_blob));
```

Comment:
fdt_totalsize() should be called only when xclmgmt_refresh_fdt_blob() returns success. I would suggest
     if (lro->core.blp_blob) {
              vfree(lro->core.blp_blob);
              lro->core.blp_blob = NULL;
    }
    ret = xclmgmt_refresh_fdt_blob()
    if (ret) return failed;
    lro->core.blp_blob = vmalloc(fdt_totalsize(lro->core.fdt_blob));

Reply:
thanks, fixed